app/src/main/kotlin/com/simplemobiletools/calculator/activities/UnitConversionActivity.kt

```diff
@@ -0,0 +1,151 @@
+    private var storedTextColor = 0
+    private var vibrateOnButtonPress = true
+    private var storedUseEnglish = false
+
+    lateinit var calc: ConversionCalculatorImpl
+
+    @SuppressLint("MissingSuperCall")
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_unit_conversion)
+
+        calc = ConversionCalculatorImpl(this, applicationContext)
+        updateViewColors(money_holder, config.textColor)
+        updateButtonColor(config.customPrimaryColor)
+
+        getButtonIds().forEach {
+            it.setOnClickListener { calc.numpadClicked(it.id); checkHaptic(it) }
+        }
+
+        result.setOnLongClickListener { copyToClipboard(result.value); true }
+        AutofitHelper.create(result)
+
+        btn_delete.setOnClickListener { calc.handleDelete(); checkHaptic(it) }
+        btn_delete.setOnLongClickListener { calc.handleClear(); true }
+
+        btn_weight.setOnClickListener { displayDialog("weight", R.layout.weight_conversion_dialog,
+                R.array.weight, R.id.from_weight_spinner, R.id.to_weight_spinner) }
+        btn_length.setOnClickListener { displayDialog("length", R.layout.length_conversion_dialog,
+                R.array.length, R.id.from_length_spinner, R.id.to_length_spinner) }
+        btn_volume.setOnClickListener { displayDialog("volume", R.layout.volume_conversion_dialog,
+                R.array.volume, R.id.from_volume_spinner, R.id.to_volume_spinner) }
+    }
+
+    @SuppressLint("MissingSuperCall")
+    override fun onResume() {
+        super.onResume()
+        if (storedUseEnglish != config.useEnglish) {
+            restartActivity()
+            return
+        }
+
+        if (storedTextColor != config.textColor) {
+            updateViewColors(money_holder, config.textColor)
+            updateButtonColor(config.customPrimaryColor)
+        }
+        vibrateOnButtonPress = config.vibrateOnButtonPress
+    }
+
+    override fun setValue(value: String) {
+        result.text = value
+    }
+
+    override fun getResult(): String {
+        return result.text.toString()
+    }
+
+    override fun setClear(text: String) {}
+    override fun getFormula(): String { return "" }
+    override fun setFormula(value: String) {}
+
+    private fun checkHaptic(view: View) {
+        if (vibrateOnButtonPress) {
+            view.performHapticFeedback()
+        }
+    }
+
+    private fun updateButtonColor(color: Int) {
+        val states = arrayOf(intArrayOf(android.R.attr.state_enabled))
+        val colors = intArrayOf(config.primaryColor)
+        val myList = ColorStateList(states, colors)
+
+        getUnitConversionButtonIds().forEach {
+            it.setTextColor(getContrastColor(color))
+            ViewCompat.setBackgroundTintList(it, myList)
+        }
+    }
+
+    @SuppressLint("Range")
+    private fun getContrastColor(color: Int): Int {
```

Comment:
Didn't wan't to restructure all the activities to inherit from a parent activity since this would have caused conflicts with a lot of the other PRs (the crypto and bitwise). We can consider doing this once all of them have been merged.

Reply:
Indeed there is a lot of duplicate code that is shared among activities so once all the features have been merged we should refactor them.

Reply:
Fair enough